avocado/utils/asset.py

```diff
@@ -371,3 +371,12 @@ def get_metadata(self):
                     metadata = json.load(f)
                     return metadata
         return None
+
+
+def find_file(name, cache_dirs, asset_hash=None, algorithm=None,
+              locations=None, expire=None):
+    """
+    Wrapper function to find a file downloaded by the Asset class in the cache.
+    """
+    return Asset(name, asset_hash, algorithm, locations, cache_dirs,
+                 expire).find_asset_file()
```

Comment:
Have you considered defining this function in the core package and having it raising a `TestCancel` exception on file not found? Calling it like `find_file(... , cancel=True)` or something like that. Just thinking out loud here.

Reply:
> Have you considered defining this function in the core package and having it raising a `TestCancel` exception on file not found? Calling it like `find_file(... , cancel=True)` or something like that. Just thinking out loud here.

The approach you are proposing is excellent, and this would decrease the complexity we throw in the user. I have not thought about that.

This week I was thinking about the opposite side. After looking at the `fetch_asset` method in the Avocado core, with a mind of a Software Engineer, it seems to me that this method should not be part of the core, as it fits on the utility class. I was ready to send an RFC to get feedback from others and also to have a historical background.

So, do you think documenting this kind of discussion would benefit the project? I can elaborate more on the subject and submit an RFC. Otherwise, if you have a straight forward answer and some historical background, shoot it here, that is fine for me.

Reply:
The `fetch_asset` method users the `self.cache_dirs` property so users don't have to pass the list of cache directories themselves. It's the "utility's client" so to say... that was designed also with a mind of a Software Engineer, I suppose.
Not sure I got your proposal though, but being the `self.fetch_asset()` a public API, I'd cautious when touching it: https://github.com/avocado-framework-tests/avocado-misc-tests/search?q=%22self.fetch_asset%22&unscoped_q=%22self.fetch_asset%22
Mind clarifying your plans there?

Reply:
> The `fetch_asset` method uses the `self.cache_dirs` property so users don't have to pass the list of cache directories themselves. 

Right, this directly benefits `find_file` and reduces the complexity of the user's side.

> It's the "utility's client" so to say... that was designed also with a mind of a Software Engineer, I suppose.

I think I was not clear about the "mind of a Software Engineer" statement. What I meant was that I'm strictly thinking about the Object Orientation concepts from the literature, where methods from a class should be actions related to the class itself. Following the theoretical concept, I don't see the activities of handling an asset as part of the core Avocado class, because I do not have the historical context from the project, so I'm sticking to the concepts here.

This does not mean we should stick with the concepts and the theory all the time. Sometimes, real-life problems in the project demand we break the general rules, but the results are worth for the users. I'm looking for the historical context of this kind of implementation to build a better knowledge about the project and to better implement the next items.

> Not sure I got your proposal though, but being the `self.fetch_asset()` a public API, I'd cautious when touching it: https://github.com/avocado-framework-tests/avocado-misc-tests/search?q=%22self.fetch_asset%22&unscoped_q=%22self.fetch_asset%22
> Mind clarifying your plans there?

The original plan was to remove all the methods that are not strictly related to the core class from it, and stick a bit more to the concepts. Still, the link you shared brought to my attention that it may not be a reasonable idea as it would have a significant impact on other projects. In this case, I don't think it is worth to have an RFC on that.

I hope I was more clear about my ideas this time, but feel free to ask if there are any blind spots.

Reply:
Let's think about the use cases here. IMO the goal of this work is to let users choose to not continue running (cancel) the test if an asset is not ready (in cache). As is, I don't see how a user would be able to keep that condition, but at the same time manually fetch the asset. Using `examples/tests/assets.py`, these are how I can see using this new API:

```python
import os

from avocado import Test
from avocado import fail_on
from avocado.utils import archive
from avocado.utils import build
from avocado.utils import process

from avocado.utils.asset import find_file


class Hello(Test):

    def setUp(self):
        mirrors = ['https://mirrors.peers.community/mirrors/gnu/hello/',
                   'https://mirrors.kernel.org/gnu/hello/',
                   'http://gnu.c3sl.ufpr.br/ftp/',
                   'ftp://ftp.funet.fi/pub/gnu/prep/hello/']
        hello = 'hello-2.9.tar.gz'
        hello_locations = ["%s/%s" % (loc, hello) for loc in mirrors]
        hello_sig = 'hello-2.9.tar.gz.sig'
        hello_sig_locations = ["%s/%s" % (loc, hello_sig) for loc in mirrors]
        self.hello = find_file(
            name=hello,
            locations=hello_locations,
            cache_dirs=self.cache_dirs)
        self.hello_sig = find_file(
            name=hello_sig,
            asset_hash='f3b9fae20c35740004ae7b8de1301836dab4ac30',
            locations=hello_sig_locations,
            cache_dirs=self.cache_dirs)

    @fail_on(process.CmdError)
    def test_gpg_signature(self):
        if not all([self.hello, self.hello_sig]):
            self.cancel('Not all assets available')
        keyring = os.path.join(self.workdir, "tempring.gpg")
        gpg_cmd = "gpg --no-default-keyring --keyring %s" % keyring
        signer_pubkey = self.get_data("gnu_hello_signer.gpg")
        import_cmd = "%s --import %s" % (gpg_cmd, signer_pubkey)
        # gpg will not return 0 when creating a new keyring and
        # importing the public key
        process.run(import_cmd, ignore_status=True)
        verify_cmd = "%s --verify %s %s" % (gpg_cmd,
                                            self.hello_sig,
                                            self.hello)
        process.run(verify_cmd)

    @fail_on(process.CmdError)
    def test_build_run(self):
        if self.hello is None:
            self.cancel('Not all assets available')
        hello_src = os.path.join(self.workdir, 'hello-2.9')
        archive.uncompress(self.hello, self.workdir)
        build.configure(hello_src)
        build.make(hello_src)
        process.run(os.path.join(hello_src, 'src', 'hello'))
```

The difference from the current version is:

```diff
diff --git a/examples/tests/assets.py b/examples/tests/assets.py
index e1a26790..63271e0d 100644
--- a/examples/tests/assets.py
+++ b/examples/tests/assets.py
@@ -6,6 +6,8 @@ from avocado.utils import archive
 from avocado.utils import build
 from avocado.utils import process
+from avocado.utils.asset import find_file

 
 class Hello(Test):
 
@@ -18,16 +20,20 @@ class Hello(Test):
         hello_locations = ["%s/%s" % (loc, hello) for loc in mirrors]
         hello_sig = 'hello-2.9.tar.gz.sig'
         hello_sig_locations = ["%s/%s" % (loc, hello_sig) for loc in mirrors]
-        self.hello = self.fetch_asset(
+        self.hello = find_file(
             name=hello,
-            locations=hello_locations)
-        self.hello_sig = self.fetch_asset(
+            locations=hello_locations,
+            cache_dirs=self.cache_dirs)
+        self.hello_sig = find_file(
             name=hello_sig,
             asset_hash='f3b9fae20c35740004ae7b8de1301836dab4ac30',
-            locations=hello_sig_locations)
+            locations=hello_sig_locations,
+            cache_dirs=self.cache_dirs)
 
     @fail_on(process.CmdError)
     def test_gpg_signature(self):
+        if not all([self.hello, self.hello_sig]):
+            self.cancel('Not all assets available')
         keyring = os.path.join(self.workdir, "tempring.gpg")
         gpg_cmd = "gpg --no-default-keyring --keyring %s" % keyring
         signer_pubkey = self.get_data("gnu_hello_signer.gpg")
@@ -42,6 +48,8 @@ class Hello(Test):
 
     @fail_on(process.CmdError)
     def test_build_run(self):
+        if self.hello is None:
+            self.cancel('Not all assets available')
         hello_src = os.path.join(self.workdir, 'hello-2.9')
         archive.uncompress(self.hello, self.workdir)
         build.configure(hello_src)

```

Which works as expected to *cancel* the tests, but then there's no way to actually fetch those assets. That would require a repetitive set of steps. Something like:

```diff
diff --git a/examples/tests/assets.py b/examples/tests/assets.py
index e1a26790..01615611 100644
--- a/examples/tests/assets.py
+++ b/examples/tests/assets.py
@@ -5,11 +5,19 @@ from avocado import fail_on
 from avocado.utils import archive
 from avocado.utils import build
 from avocado.utils import process
+from avocado.utils.asset import find_file
+
+
+# Wether to cancel test if assets are not ready (in cache)
+CANCEL_IF_ASSETS_NOT_READY = os.environ.get('CANCEL_IF_ASSETS_NOT_READY', False)
 
 
 class Hello(Test):
 
     def setUp(self):
+        self._check_or_fetch_assets(CANCEL_IF_ASSETS_NOT_READY)
+
+    def _check_or_fetch_assets(self, check_only=False):
         mirrors = ['https://mirrors.peers.community/mirrors/gnu/hello/',
                    'https://mirrors.kernel.org/gnu/hello/',
                    'http://gnu.c3sl.ufpr.br/ftp/',
@@ -17,14 +25,29 @@ class Hello(Test):
         hello = 'hello-2.9.tar.gz'
         hello_locations = ["%s/%s" % (loc, hello) for loc in mirrors]
         hello_sig = 'hello-2.9.tar.gz.sig'
+        hello_sig_hash='f3b9fae20c35740004ae7b8de1301836dab4ac30'
         hello_sig_locations = ["%s/%s" % (loc, hello_sig) for loc in mirrors]
-        self.hello = self.fetch_asset(
-            name=hello,
-            locations=hello_locations)
-        self.hello_sig = self.fetch_asset(
-            name=hello_sig,
-            asset_hash='f3b9fae20c35740004ae7b8de1301836dab4ac30',
-            locations=hello_sig_locations)
+
+        if check_only:
+            self.hello = find_file(
+                name=hello,
+                locations=hello_locations,
+                cache_dirs=self.cache_dirs)
+            self.hello_sig = find_file(
+                name=hello_sig,
+                asset_hash=hello_sig_hash,
+                locations=hello_sig_locations,
+                cache_dirs=self.cache_dirs)
+            if not (self.hello and self.hello_sig):
+                self.cancel('Not all assets available')
+        else:
+            self.hello = self.fetch_asset(
+                name=hello,
+                locations=hello_locations)
+            self.hello_sig = self.fetch_asset(
+                name=hello_sig,
+                asset_hash=hello_sig_hash,
+                locations=hello_sig_locations)
 
     @fail_on(process.CmdError)
     def test_gpg_signature(self):

```

So, because of that, I lean towards @apahim suggestion of providing an automated action when a given condition (file not found on cache) happens. But, I'd try to avoid introducing another core method, because:
 * the parameters would be very similar to the existing `fetch_asset`
 * keeping backwards compatibility for the core methods are of very high importance, so it's better to avoid introducing new ones

So, what if we just beef up `fetch_asset` with:
 * `only_if_ready` (or similar) parameter that will cause a condition to be executed
 * the condition for now can simply be a default one that is getting the test canceled
 * (in the future, maybe) a `not_ready_action` parameter that will execute some other action

Reply:
@clebergnu thanks for the examples. The use case I was thinking is much more limited and relates to the avocado + qemu acceptance tests. The use case I had in mind is just a matter of running an `avocado assets fetch <acceptance test>` to fetch the assets and using `find_file` as the first thing in the test. The `asset_fetch` can come later in the same code, and the verification of whether the test will run or not if an asset is not available can happen with an environment variable.

This PR was a straight forward implementation to make the function available to be used as soon as possible without significant impacts to the current code. The complex implementation will come later, with the introduction of the _Requirements Resolver_.

Anyway, I'm open to do an implementation of a more complex and broader solution right now.

Reply:
What I don't understand is what happens if a QEMU acceptance test is run *without* the `avocado assets fetch` command prior to it.  Or even with `avocado assets fetch` command being run, will it actually fetch any assets without a explicit `self.fetch_asset()` call?

Maybe you can provide an example test using `find_file` that will actually let me optionally fetch the assets (that is, optionally let me cancel the test if the asset is not available) while at the same time allowing me to download them both at `avocado assets fetch` time, and at `avocado run` time.

Reply:
That is actually what I meant with "The asset_fetch can come later in the same code". We will still need a _find_ + _fetch_ block. Like:

```
asset_file = find_file(...)
if asset_file is None:
    if self.cancel_if_asset_not_available:
        self.cancel("Missing Asset")
    else:
        self.fetch_asset(...)
```

Personally, I have difficulty thinking about an architecture that implements the _find_ and the _fetch_ commands in one single function. They are totally different actions. But, again, I'm open to suggestions and if we conclude that it may be the best for the users and the project, I'm in.

Reply:
Anyway, I'm going to work on a new PR to extend the current `fetch_asset` method in the Avocado core to handle the possibility of canceling the test if the asset is not available in the cache and the user do not want to download it.

I'll keep this PR opened and we can decide which solution is preferred after both are available.

Reply:
Yep, something like `self.fetch_asset(name, cache_only=True, cancel_on_missing=True)` seems like a reasonable implementation for the use case & for keeping the current user experience and interface.